rln-research/membership-merkle-tree-storage-evaluation.md

```diff
@@ -0,0 +1,107 @@
+# Waku-RLN-Relay: Evaluating Storage Overhead of Membership Merkle Tree   
+## Introduction
+
+In the current design of the Waku-RLN-Relay protocol, the membership Merkle tree (MT) is locally persisted by each peer. This incurs some storage overhead which may or may not fit resource-limited devices. This overhead also depends on the algorithm used to represent the Merkle tree. So far, we are aware of two algorithms namely, Full Merkle tree (FMT) and Optimal Merkle tree (OMT), and in this post, we would like
+1) to compare the two existing MT algorithms and clarify their differences in terms of storage requirements.
+2) to determine whether the storage overhead of the membership MT fits the resource-limited devices. 
+3) depending on the second item, then we can decide on whether more research on the MT storage optimization is required or not.
+
+**Existing Merkle tree algorithms**: In the present post, we analyze two different MT algorithms:
+- Full Merkle tree (FMT): in this algorithm, an empty full tree with a specific depth is being initialized and the required memory space is allocated upfront. The values of the existing nodes are altered as new nodes are added to the tree.
+- Optimal Merkle tree (OMT): In this algorithm, which is implemented in the  [killic library](https://github.com/status-im/rln/blob/master/src/merkle.rs), the Merkle tree nodes are generated incrementally as new leaves are being added to the tree.
+
+## Storage Overhead Analysis
+ 
+**Per-app memory cap**: Before starting the analysis, it is worth reviewing the per-app memory cap on mobile devices (Android and iOS) which are more resource-restricted compared to desktop nodes. Note that these values are not accurate and not evaluated by us, and merely to give some intuition on per-app memory contraints:
```

Comment:
Don't have to go into too much detail on this here, but brower also has some relevant restrictions.

Reply:
It seems the memory restriction of the browsers does impose real limitation for us: 
> The latest version of Google Chrome (desktop 64-bit version), ver 80.0, can handle up to 4GB of memory per tab as of April 2020

Reply:
Hum, I tried up to 3.5GB with Chromium using https://gist.github.com/fryorcraken/0bbfe69c0eb947e655707b87ccf3ff14

But then you said 100MB in Discord. Do you need more than 4GB?

Reply:
Thanks @fryorcraken for experimenting!  3.5 GB is beyond our need, our requirement is around `100` MB. So it means that we won't have an issue loading the membership Merkle tree in a browser.